src/xopen/__init__.py

```diff
@@ -356,6 +356,46 @@ def _open_gz(filename, mode, compresslevel, threads):
         return buffered_writer(gzip.open(filename, mode, compresslevel=compresslevel))
 
 
+def _detect_format_from_content(filename, mode, compresslevel, threads):
```

Comment:
It’s great you factored out this function, but what I initially thought was to separate detecting file contents from opening the file (I should have been clearer about this, sorry): So `_detect_format_from_content` and `_detect_format_from_name` would actually only do the detection work and return the file type (perhaps as a string `bz2`, `xz`, `gz` or None), and then within the `xopen` function itself, this result gets used to use the correct function to open the file. As it is now, the `_detect` functions also do the opening, so I’d say that at least their names aren’t quite accurate.

And regarding the exception, I think it would be fine to catch the `IOError` and just return None as "detection result" (for "unknown"). Detecting the format from the name works like this as well: If the name isn’t recognized, it just falls through to a regular `open()`.

Reply:
Oh, I think you communicated what you wanted perfectly. I was trying to be clever and avoid opening the file, if possible, while keeping the number of conditionals to a minimum. 

But, would something along the lines the code below work better?
```python
 if filename == '-':
     return _open_stdin_or_out(mode)
detected_format = _detect_format_from_content(filename, mode)
if filename.endswith('.bz2') or detected_format == "bz2":
     return _open_bz2(filename, mode)
elif filename.endswith('.xz') or detected_format == "xz":
     return _open_xz(filename, mode)
elif filename.endswith('.gz') or detected_format == "gz":
     return _open_gz(filename, mode, compresslevel, threads)
else:
     # Python 2.6 and 2.7 have io.open, which we could use to make the returned
     # object consistent with the one returned in Python 3, but reading a file
     # with io.open() is 100 times slower (!) on Python 2.6, and still about
     # three times slower on Python 2.7 (tested with "for _ in io.open(path): pass")
     return open(filename, mode) 
```
 If so, I'll make a commit with those changes.

Reply:
It’s not super important, but I’d suggest something like this (untested):
```python
if filename == '-':
     return _open_stdin_or_out(mode)
detected_format = _detect_format_from_extension(filename)
if detected_format is None and "w" not in mode:
    detected_format = _detect_format_from_content(filename)

if detected_format == "bz2":
    return _open_bz2(filename, mode)
elif detected_format == "xz":
    return _open_xz(filename, mode)
elif detected_format == "gz":
    return _open_gz(filename, mode, compresslevel, threads)
else:
    return open(filename, mode)
```
Then the first five lines deal with detecting the file format, and the rest deals with opening the file according to the detected format. It separates the concerns (detection vs opening) a bit better.

But this is essentially nitpicking. I just read the book *Clean Code*, which made me pay more attention to these types of thing :-).

Reply:
Thanks for the feedback. That looks good. I added a commit with those changes. Nitpicking is okay, it is your project after all!